File: mirror/tier2-python/showcase/async_promises.py
"""
MIRROR V3: Tier 2 Python Showcase
Category: Async & Control Flow
Module: async_promises.py
Purpose: Demonstrate promise-like patterns
"""

from typing import Callable, Any, List


class Promise:
    """Promise-like class"""
    def __init__(self, executor: Callable):
        self.state = "pending"
        self.value = None
        self.callbacks: List[Callable] = []
        
        def resolve(result):
            if self.state == "pending":
                self.state = "fulfilled"
                self.value = result
                for callback in self.callbacks:
                    callback(result)
        
        executor(resolve)
    
    def then(self, callback: Callable) -> 'Promise':
        """Add callback"""
        if self.state == "fulfilled":
            callback(self.value)
        else:
            self.callbacks.append(callback)
        return self
    
    def get_value(self):
        """Get promise value"""
        return self.value


def create_promise_from_value(value):
    """Create resolved promise"""
    return Promise(lambda resolve: resolve(value))


p1 = Promise(lambda resolve: resolve(42))
p2 = Promise(lambda resolve: resolve("hello"))

result = {
    "promise_state": p1.state,
    "promise_value": p1.get_value(),
    "promise_2_value": p2.get_value(),
    "promise_created": True,
}

if __name__ == "__main__":
    print(result)

File: mirror/tier2-php/showcase/async_promises.php
<?php
// MIRROR V3: Tier 2 PHP Showcase
// Category: Async & Control Flow
// Module: async_promises.php

class Promise {
    private $state = 'pending';
    private $value = null;
    private $callbacks = [];
    
    public function __construct(callable $executor) {
        $executor(function($result) {
            if ($this->state === 'pending') {
                $this->state = 'fulfilled';
                $this->value = $result;
                foreach ($this->callbacks as $callback) {
                    $callback($result);
                }
            }
        });
    }
    
    public function then(callable $callback): self {
        if ($this->state === 'fulfilled') {
            $callback($this->value);
        } else {
            $this->callbacks[] = $callback;
        }
        return $this;
    }
    
    public function getValue() {
        return $this->value;
    }
    
    public function getState(): string {
        return $this->state;
    }
}

$p1 = new Promise(fn($resolve) => $resolve(42));
$p2 = new Promise(fn($resolve) => $resolve('hello'));

$result = [
    'promise_state' => $p1->getState(),
    'promise_value' => $p1->getValue(),
    'promise_2_value' => $p2->getValue(),
    'promise_created' => true,
];

echo json_encode($result);
?>
